File: backend/api/sessions.py
from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from db.engine import get_db
from db.repositories import messages as messages_repo
from db.repositories import sessions as sessions_repo

router = APIRouter()


class CreateSessionRequest(BaseModel):
    title: str | None = None


class SessionResponse(BaseModel):
    id: UUID
    title: str
    created_at: datetime

    model_config = {"from_attributes": True}


class MessageResponse(BaseModel):
    id: UUID
    role: str
    content: str
    created_at: datetime

    model_config = {"from_attributes": True}


@router.post("/sessions", response_model=SessionResponse)
async def create_session(body: CreateSessionRequest, db: AsyncSession = Depends(get_db)):
    session = await sessions_repo.create_session(db, title=body.title or "New Session")
    return session


@router.get("/sessions/{session_id}/history", response_model=list[MessageResponse])
async def get_history(session_id: UUID, db: AsyncSession = Depends(get_db)):
    session = await sessions_repo.get_session(db, session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return await messages_repo.get_history(db, session_id)

File: node-server/src/api/sessions.ts
import type { FastifyPluginAsync } from "fastify";
import { z } from "zod";
import { requireAuth } from "../auth/plugin.js";
import { db } from "../db/client.js";
import * as messagesRepo from "../db/repositories/messages.js";
import * as sessionsRepo from "../db/repositories/sessions.js";

const createSessionBody = z.object({
  title: z.string().nullish(),
});

const sessionIdParams = z.object({
  session_id: z.string().uuid(),
});

export const sessionsRoutes: FastifyPluginAsync = async (app) => {
  app.post("/sessions", { preHandler: requireAuth }, async (request, reply) => {
    const parsed = createSessionBody.safeParse(request.body ?? {});
    if (!parsed.success) {
      return reply.code(422).send({ detail: parsed.error.issues });
    }

    // Owner comes from the verified cookie, never from the request body.
    const session = await sessionsRepo.createSession(
      db,
      request.user!.id,
      parsed.data.title || "New Session",
    );

    return {
      id: session.id,
      title: session.title,
      created_at: session.createdAt.toISOString(),
    };
  });

  app.get("/sessions/:session_id/history", { preHandler: requireAuth }, async (request, reply) => {
    const parsed = sessionIdParams.safeParse(request.params);
    if (!parsed.success) {
      return reply.code(422).send({ detail: parsed.error.issues });
    }
    const sessionId = parsed.data.session_id;

    // 404 rather than 403 for someone else's session — a 403 would confirm the
    // id exists.
    const session = await sessionsRepo.getSessionForUser(db, sessionId, request.user!.id);
    if (!session) {
      return reply.code(404).send({ detail: "Session not found" });
    }

    const history = await messagesRepo.getHistory(db, sessionId);

    return history.map((m) => ({
      id: m.id,
      role: m.role,
      content: m.content,
      created_at: m.createdAt.toISOString(),
    }));
  });
};
